{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4a9e50fc-9ff0-4032-a00b-26887308987c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1834685e-cd54-4f41-ada6-970b2fe75607",
   "metadata": {},
   "source": [
    "This script calculates the symmetric SNP differences between two groups of genomes. The output consists of the minimum, mean and maximum as well as a dataframe of all pairwise symmetric SNP differences. The input file \"annotated_variants.tab\" can be generated using the perSVade pipeline\n",
    "The script can also compare two individual genomes or 1 genome vs a group of genomes\n",
    "\n",
    "It is recommended that each \"annotated_variants.tab\" file is stored in a directory named after the genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "192b97af-1ddf-4cd6-931f-ada108aed5c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "path1 = input(\"Provide the path to the first group of annotated variants:\")\n",
    "path2 = input(\"Provide the path to the second group of annotated variants:\")\n",
    "save_matrix = input(\"Do you with to save the matrix? (y/n):\")\n",
    "\n",
    "dir_list1 = os.listdir(path1)\n",
    "dir_list2 = os.listdir(path2)\n",
    "\n",
    "def import_command(dir_list, path):\n",
    "    set_dict = {}\n",
    "    for item in dir_list:\n",
    "        if item in dir_list:\n",
    "            try:\n",
    "                cp = pd.read_csv(path + \"/\"+ item + \"/annotated_variants.tab\", delimiter=\"\\t\")\n",
    "                cp = cp[cp['is_snp'] == True]\n",
    "                set_dict[item] = set(cp['#Uploaded_variation'])\n",
    "            except Exception:\n",
    "                pass\n",
    "    return set_dict\n",
    "\n",
    "set_dict1 = import_command(dir_list1, path1)\n",
    "set_dict2 = import_command(dir_list2, path2)\n",
    "\n",
    "dict1 = {}\n",
    "for key1, value1 in set_dict1.items():\n",
    "    dict2= {}\n",
    "    for key2, value2 in set_dict2.items():\n",
    "        dict2[key2] = len(value1.symmetric_difference(value2))\n",
    "    dict1[key1] = dict2\n",
    "\n",
    "df = pd.json_normalize([{'index': k, **v} for k, v in dict1.items()])\n",
    "df = df.set_index('index')\n",
    "if save_matrix == \"y\":\n",
    "    save_location = input(\"Where do you want to save the matrix?:\")\n",
    "    df.to_csv(save_location + \"/matrix.csv\")\n",
    "else:\n",
    "    print(\"Matrix was not saved\")\n",
    "print(\"The minimum symmetric SNP-difference is: \" + str(df.min().min()))\n",
    "print(\"The rounded mean symmetric SNP-difference is: \" + str(int(round(df.mean().mean(), 0))))\n",
    "print(\"The maximum symmetric SNP-difference is: \" + str(df.max().max()))\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "580567f0-ea11-4331-b591-00737bb5467b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
